File: 2022/haskell/src/Day01.hs
module Main where


-- import

import Data.List       (sort)
import Data.List.Split

import AOC


-- input

type Input = [Int]

parseInput :: String -> Input
parseInput = map (sum . map read) . splitOn [""] . lines


-- solution

part1 :: Input -> Int
part1 = maximum

part2 :: Input -> Int
part2 = sum . take 3 . reverse . sort


-- main

main :: IO ()
main = aocMain 01 $ \rawData -> do
  let testInput = parseInput example
      realInput = parseInput rawData
  putStrLn "# Part 1"
  print $ part1 testInput
  print $ part1 realInput
  putStrLn "# Part 2"
  print $ part2 testInput
  print $ part2 realInput

example :: String
example = "1000\n2000\n3000\n\n4000\n\n5000\n6000\n\n7000\n8000\n9000\n\n10000"

File: 2022/cpp/src/day01.cc
#include <vector>
#include <iostream>
#include <algorithm>

#include "aoc.hh"

const std::vector<std::string> example =
  { "1000",
    "2000",
    "3000",
    "",
    "4000",
    "",
    "5000",
    "6000",
    "",
    "7000",
    "8000",
    "9000",
    "",
    "10000"
  };

void part1(const std::vector<std::string>& lines)
{
  std::vector<int> calories;
  int current = 0;

  for (auto const& line : lines)
  {
    if (line.empty())
    {
      calories.push_back(current);
      current = 0;
      continue;
    }
    current += std::stoi(line);
  }
  calories.push_back(current);

  std::cout << "part1: " << *std::max_element(calories.begin(), calories.end()) << std::endl;
}

int main()
{
  auto input = aoc::get_input_lines(1);
  part1(example);
  part1(input);
}
